### PR TITLE
Correctly track train / validation losses

### LINKED ISSUE #481: BasicWorkflow: Unstable val_losses due to validation data shuffling
In the `BasicWorkflow` class, the validation set is constructed via `validation_data = OfflineDataset(data=validation_data, batch_size=dataset.batch_size, adapter=self.adapter)` [in the `_fit` method](https://github.com/bayesflow-org/bayesflow/blob/main/bayesflow/workflows/basic_workflow.py#L882). This treats the validation data as a standard `OfflineDataset` and implies shuffling on epoch end. Thus, stochasticity is introduced by modifying the composition of the evaluated batches, which can lead to substantial changes between multiple validation loss calculations during training even for frozen networks. 

E.g., adapting Linear_Regression_Starter.ipynb to use offline training with a learning rate of 0 (and no standardization in the adapter),

```python
validation_data = workflow.simulate(1000)
history = workflow.fit_offline(
    data=training_data,
    epochs=5,
    batch_size=32,
    validation_data=validation_data,
)
```

leads to highly variable validation losses:

```python
[19.465839385986328,
 10.662302017211914,
 16.612285614013672,
 7.247644901275635,
 17.282154083251953]
```

Setting the number of validation sets equal the batch size, either via `validation_data = workflow.simulate(32)` or via `batch_size=1000`, enforces equal batches despite shuffling, resulting in stable validation losses:

```python
[12.364879608154297,
 12.364879608154297,
 12.364879608154297,
 12.364879608154297,
 12.364879608154297]
```

A pull request with a simple fix will follow.

### PR DESCRIPTION
**Pull request by LarsKue**
This PR:

- Fixes #481 
- Supercedes #483 
- Fixes validation loss tracking in `test_step`
- Removes `loss/..._loss` and other nested metrics
- Temporarily removes the ability to track custom metrics

Future TODO:
- We have to reenable tracking custom metrics by keeping a `keras.metrics.Mean` tracker object on the approximator for each custom metric the user passes. We also need to call `update_state` on each of those trackers in the `compute_metrics` or `train/test_step` method.

@vpratz Do you think you have capacity to take care of the TODO?

### CODE REVIEW

**Comment by vpratz** [comment]
Do I read it correctly that this PR does not modify the shuffling and therefore not fix this aspect of #481?

Regarding the TODO: I'm not familiar with this part of the code base yet, so I cannot really judge how complicated it is and when I will find the time, but I can try to a closer look at it in the coming weeks.

**Comment by stefanradev93** [comment]
Shall we still merge this PR, as it provides a pretty important fix and open another one pertaining to general metrics and aggregation.

**Comment by stefanradev93** [comment]
It fixes the incorrect calculation and display of validation losses only on the last validation batch. Additionally, it removes the duplicate printing of loss, inference_net/loss, val_loss / val_inference_net/loss.

**Comment by vpratz** [comment]
Ahh, thanks a lot! I will do a proper review in the next days, then.

**Comment by vpratz** [comment]
I now understand better what is going on, and the changes look good to me. I will try to add the trackers for the other metrics today, and let you know if I encounter any difficulties...

**Comment by vpratz** [comment]
Quick question @LarsKue @stefanradev93 : The individual losses were removed in the PR. I would add them as metrics in the case that more than one loss is present (not if there is only one, to avoid the useless duplicate), so that they are individually tracked. This will also display them, which I think is desirable. Do you agree, or would you rather not display that information?

**Comment by vpratz** [comment]
Short update: I have succeeded in tracking the metrics, but the serialization for custom `inference_metrics` and `summary_metrics` seems to be broken (at least in TensorFlow), and might need some restructuring to work nicely. It might make sense to move this to a separate PR though, I'm currently checking if the issues are separate or interdependent...

**Comment by vpratz** [comment]
@LarsKue @stefanradev93 I think the changes are ready to review, the issues regarding custom metrics are (as far as I can tell) not related to the changes in this PR.

**Comment by vpratz** [comment]
I was able to resolve this properly. The reason for the order-dependency was taking the unweighted mean of means. So in the averaged metrics, average values obtained from batches with batch size had the same weight as values obtained from batch size 2.
By setting the `sample_weight` argument of `Metric.update_state` to the batch size, we can obtain the correct averages. This requires calculating the batch sizes from the data at hand, which is approximator-specific. Therefore I outsourced this calculation into a private method that each approximator has to overwrite. If you have different design-ideas for this, please let me know.

This fixes #481 and supersedes that aspect of #482.
@LarsKue @stefanradev93 Could you please take another look?

Note: `tqdm` will report wrong values for the training loss, as it does naive sampling, the value stored in `history` will be correct, though.

<details><summary>Code to reproduce</summary>

```python
import bayesflow as bf
import keras

workflow = bf.BasicWorkflow(
    inference_network=bf.networks.CouplingFlow(subnet_kwargs={"dropout": 0.0}),
    inference_variables=["parameters"],
    inference_conditions=["observables"],
    simulator=bf.simulators.GaussianMixture(),
    initial_learning_rate=0.0,
    standardize=[]
)

training_data = workflow.simulate(66)
validation_data = workflow.simulate(66)
history = workflow.fit_offline(
    data=training_data,
    epochs=3,
    batch_size=32,
    validation_data=validation_data,
)
print(history.history)
```

</details>

**Review by stefanradev93 — COMMENTED** [review]
Hi Valentin and Lars, I buffed up the docs for the new changes. I also decided to repeat `_batch_size_from_data` in the backend-specific classes to avoid confusion, since the method is already explicitly needed in these classes. Other than that, I think this PR is ready to merge.
